src/Filters/buttord.jl

```diff
@@ -0,0 +1,190 @@
+    # NOTE: the optimization function will adjust the corner frequencies in Wp, returning a new passband tuple.
+    Δ = eps(typeof(Wp[1]))^(2/3)
+
+    # optimize order on bound [passband low < w < stopband-tolerance].
+    C₁(w) = bsfcost(w, true, Wp, Ws, Rp, Rs)
+    p1 = minimizer(optimize(C₁, Wp[1], Ws[1]-Δ))
+    
+    # declaring the 2nd cost function here to use the new passband tuple.
+    C₂(w) = bsfcost(w, false, tuple(p1, Wp[2]), Ws, Rp, Rs)
+    p2 = minimizer(optimize(C₂, Ws[2]+Δ, Wp[2]))
+    Wadj = tuple(p1, p2)
+
+    Wa = (Ws .* (Wadj[1]-Wadj[2])) ./ (Ws.^2 .- (Wadj[1]*Wadj[2]))
+    minimum(abs.(Wa)), Wadj
+end
+
+fromprototype(Wp::Real, Wscale::Real, ftype::Type{Lowpass}) = Wp * Wscale
+fromprototype(Wp::Real, Wscale::Real, ftype::Type{Highpass}) = Wp / Wscale
+
+function fromprototype(Wp::Tuple{Real,Real}, Wscale::Real, ftype::Type{Bandstop})
+    Wa = zeros(2)
+    diff = Wp[2]-Wp[1]
+    prod = Wp[2]*Wp[1]
+    Wa[1] = (diff + sqrt(diff^2 + 4*(Wscale^2)*prod)) / (2*Wscale)
+    Wa[2] = (diff - sqrt(diff^2 + 4*(Wscale^2)*prod)) / (2*Wscale)
+    sort!(abs.(Wa))
+end
+
+function fromprototype(Wp::Tuple{Real,Real}, Wscale::Real, ftype::Type{Bandpass})
+    Wsc = [-Wscale, Wscale]
+    Wa = -Wsc .* (Wp[2]-Wp[1])./2 .+ sqrt.( Wsc.^2/4*(Wp[2]-Wp[1]).^2 .+ Wp[1]*Wp[2])
+    sort!(abs.(Wa))
+end
+
+order_estimate(Rp::Real, Rs::Real, warp::Real) = (log10(db2pow(Rs) - 1) - log10(db2pow(Rp) - 1)) / (2*log10(warp))
+natfreq_estimate(warp::Real, Rs::Real, order::Integer) = warp / (^(10, 0.1*Rs) - 1)^(1/(2*order))
+
+
+"""
+    (N, ωn) = buttord(Wp, Ws, Rp, Rs)
+
+Butterworth order estimate for bandpass and bandstop filter types. 
+`Wp` and `Ws` are 2-element pass and stopband frequency edges, with 
+no more than `Rp` dB passband ripple and at least `Rs` dB stopband 
+attenuation. Based on the ordering of passband and bandstop edges, 
+the Bandstop or Bandpass filter type is inferred. `N` is an integer 
+indicating the lowest estimated filter order, with `ωn` specifying
+the cutoff or "-3 dB" frequencies.
+"""
+function buttord(Wp::Tuple{Real,Real}, Ws::Tuple{Real,Real}, Rp::Real, Rs::Real)
+
+    # make sure the band edges are in increasing order.
+    Wps = (Wp[1] > Wp[2]) ? tuple(Wp[2], Wp[1]) : tuple(Wp[1], Wp[2])
+    Wss = (Ws[1] > Ws[2]) ? tuple(Ws[2], Ws[1]) : tuple(Ws[1], Ws[2])
+
+    # infer filter type based on ordering of edges.
+    ftype = (Wps[1] < Wss[1]) ? Bandstop : Bandpass
+
+    # pre-warp both components
+    wp = tan.(π/2 .* Wps)
+    ws = tan.(π/2 .* Wss)
+
+    if (ftype == Bandstop)
+        # optimizer modifies passband frequencies.
+        (wa, wpadj) = toprototype(wp, ws, Rp, Rs, ftype)
+    else
+        # no optimization in BPF case, use original warped passband edges.
+        wa = toprototype(wp, ws, ftype)
+        wpadj = wp
+    end
+
+    # get the integer order estimate.
+    N = ceil(Int, order_estimate(Rp, Rs, wa))
+
+    wscale = natfreq_estimate(wa, Rs, Integer(N))
```

Comment:
`N` is now an  `Int`, so
```suggestion
    wscale = natfreq_estimate(wa, Rs, N)
```